compiled failed on windows

**Describe the bug**
Can't compile CHOLMOD

**To Reproduce**
mkdir build
cd build
`cmake ..  -DCMAKE_CXX_COMPILER=icc -DCMAKE_C_COMPILER=icc -DCMAKE_Fortran_COMPILER=ifort -DCMAKE_INSTALL_PREFIX="E:/umfpackCompilerWork/SuiteSparse7.7" -DSUITESPARSE_ENABLE_PROJECTS="umfpack"`
the result of cmake stored in `config.log`
[config.log](https://github.com/DrTimothyAldenDavis/SuiteSparse/files/15258994/config.log)
`msbuild SuiteSparse.sln /p:Configuration=Release > compile.log`
the result of build stored in compile.log
[compile.log](https://github.com/DrTimothyAldenDavis/SuiteSparse/files/15258993/compile.log)

**Expected behavior**
normal compilation

**Screenshots**
![image](https://github.com/DrTimothyAldenDavis/SuiteSparse/assets/44524468/cd13d583-5736-4ae0-b5d8-d23a9cd8218f)


**Desktop (please complete the following information):**
 - OS: Windows 11
 - compiler MSVC
 - intel mkl
 - Version 7.7.0




**Additional context**
I found `-DCMAKE_CXX_COMPILER=icc -DCMAKE_C_COMPILER=icc` also not work.

Found a direct download link for the Intel Fortran compilers:
https://www.intel.com/content/www/us/en/developer/articles/tool/oneapi-standalone-components.html#fortran

That installed `ifort`:
```
Intel(R) Fortran Intel(R) 64 Compiler Classic for applications running on Intel(R) 64, Version 2021.12.0 Build 20240222_000000
Copyright (C) 1985-2024 Intel Corporation.  All rights reserved.

ifort: remark #10448: Intel(R) Fortran Compiler Classic (ifort) is now deprecated and will be discontinued late 2024. Intel recommends that customers transition now to using the LLVM-based Intel(R) Fortran Compiler (ifx) for continued Windows* and Linux* support, new language support, new language features, and optimizations. Use '/Qdiag-disable:10448' to disable this message.
```

With it, compilation fails for me (but with a different error than what you were showing). I'm using the following commands to configure and build:
```
cmake -G"Ninja Multi-Config" -DCMAKE_CXX_COMPILER=icx-cl -DCMAKE_C_COMPILER=icx-cl -DCMAKE_Fortran_COMPILER=ifort -DSUITESPARSE_ENABLE_PROJECTS="umfpack" --fresh ..
cmake --build . --config Release
```
The error is on linking `cholmod.dll`:
```
[630/1250] Linking CUDA shared library CHOLMOD\Release\cholmod.dll
FAILED: CHOLMOD/Release/cholmod.dll CHOLMOD/Release/cholmod.lib
C:\WINDOWS\system32\cmd.exe /C "C:\WINDOWS\system32\cmd.exe /C ""C:\Program Files\Microsoft Visual Studio\2022\Community\Common7\IDE\CommonExtensions\Microsoft\CMake\CMake\bin\cmake.exe" -E __create_def D:\repo\SuiteSparse\SuiteSparse\build-oneAPI\CHOLMOD\CMakeFiles\CHOLMOD.dir\Release\exports.def D:\repo\SuiteSparse\SuiteSparse\build-oneAPI\CHOLMOD\CMakeFiles\CHOLMOD.dir\Release\exports.def.objs && cd D:\repo\SuiteSparse\SuiteSparse\build-oneAPI" && "C:\Program Files\Microsoft Visual Studio\2022\Community\Common7\IDE\CommonExtensions\Microsoft\CMake\CMake\bin\cmake.exe" -E vs_link_dll --intdir=CHOLMOD\CMakeFiles\CHOLMOD.dir\Release --rc=C:\PROGRA~2\WI3CF2~1\10\bin\100226~1.0\x64\rc.exe --mt=C:\PROGRA~2\WI3CF2~1\10\bin\100226~1.0\x64\mt.exe --manifests  -- C:\PROGRA~1\MIB055~1\2022\COMMUN~1\VC\Tools\MSVC\1439~1.335\bin\Hostx64\x64\link.exe /nologo @CMakeFiles\CHOLMOD.Release.rsp  /out:CHOLMOD\Release\cholmod.dll /implib:CHOLMOD\Release\cholmod.lib /pdb:CHOLMOD\Release\cholmod.pdb /dll /version:5.2 /Qoption,link,/machine:x64 /INCREMENTAL:NO  -Qiopenmp /Qoption,link,/DEF:CHOLMOD\CMakeFiles\CHOLMOD.dir\Release\exports.def -LIBPATH:"C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v12.3/lib/x64"  && cd ."
LINK: command "C:\PROGRA~1\MIB055~1\2022\COMMUN~1\VC\Tools\MSVC\1439~1.335\bin\Hostx64\x64\link.exe /nologo @CMakeFiles\CHOLMOD.Release.rsp /out:CHOLMOD\Release\cholmod.dll /implib:CHOLMOD\Release\cholmod.lib /pdb:CHOLMOD\Release\cholmod.pdb /dll /version:5.2 /Qoption,link,/machine:x64 /INCREMENTAL:NO -Qiopenmp /Qoption,link,/DEF:CHOLMOD\CMakeFiles\CHOLMOD.dir\Release\exports.def -LIBPATH:C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v12.3/lib/x64 /MANIFEST:EMBED,ID=2" failed (exit code 1181) with the following output:
LINK : warning LNK4044: unrecognized option '/Qoption,link,/machine:x64'; ignored
LINK : warning LNK4044: unrecognized option '/Qiopenmp'; ignored
LINK : warning LNK4044: unrecognized option '/Qoption,link,/DEF:CHOLMOD\CMakeFiles\CHOLMOD.dir\Release\exports.def'; ignored
LINK : fatal error LNK1181: cannot open input file 'AMD\Release\amd.lib'
```
And there is indeed no `amd.lib` (just `amd.dll` and `amd_static.lib`). It looks like that `amd.dll` doesn't export any symbols...

It correctly exports the symbols from `amd.dll` using the (modern, LLVM-based) Intel Fortran compiler `ifx`:
```
cmake -DCMAKE_CXX_COMPILER=icx-cl -DCMAKE_C_COMPILER=icx-cl -DCMAKE_Fortran_COMPILER=ifx -DSUITESPARSE_ENABLE_PROJECTS="umfpack" --fresh ..
cmake --build . --config Release
```

However, using `ifx` still fails a bit later on when linking the CHOLMOD demos. That seems to be caused by `cholmod.dll` exporting only one single symbol:
```
>dumpbin /EXPORTS CHOLMOD\Release\cholmod.dll
Microsoft (R) COFF/PE Dumper Version 14.39.33523.0
Copyright (C) Microsoft Corporation.  All rights reserved.


Dump of file CHOLMOD\Release\cholmod.dll

File Type: DLL

  Section contains the following exports for cholmod.dll

    00000000 characteristics
    FFFFFFFF time date stamp
        0.00 version
           1 ordinal base
           1 number of functions
           1 number of names

    ordinal hint RVA      name

          1    0 0003FCEC NvOptimusEnablementCuda

  Summary

        9000 .data
        1000 .nvFatBi
        9000 .nv_fatb
        1000 .pdata
       1D000 .rdata
        2000 .reloc
        1000 .rsrc
       1D000 .text
        2000 _RDATA
        1000 __nv_mod
        4000 __nv_rel
```

Afaict, that is because `/Qoption,link,/DEF:CHOLMOD\CMakeFiles\CHOLMOD.dir\Release\exports.def` is used in the `LINK_FLAGS` when linking `cholmod.dll` as a CUDA library. That should be `/DEF:CHOLMOD\CMakeFiles\CHOLMOD.dir\Release\exports.def` instead (without the prefix `/Qoption,link,`).

The same issue with `amd.dll` and `ifort`. When linking as a Fortran library with `ifort`, the `LINK_FLAGS` shouldn't contain `/Qoption,link,/DEF:AMD\CMakeFiles\AMD.dir\Release\exports.def`. It should be `/DEF:AMD\CMakeFiles\AMD.dir\Release\exports.def` instead when using the (non LLVM-based) `ifort` (which leads to CMake calling `link.exe` directly like for the CUDA library).

That looks like a CMake issue to me. There is probably nothing that SuiteSparse can do about it.

Disabling CUDA leads to `cholmod.dll` being linked as a C library. That can be used to work around the issue until this is fixed in CMake.
Similarly, disabling Fortran leads to `amd.dll` being linked as a C library. That (or using the LLVM-based `ifx`) can be used to work around the issue with the empty `amd.dll`.

I didn't find an open ticket for either issue on https://gitlab.kitware.com/cmake/cmake/-/issues.

--- a/AMD/CMakeLists.txt
+++ b/AMD/CMakeLists.txt
@@ -90,7 +90,8 @@ if ( BUILD_SHARED_LIBS )
         OUTPUT_NAME amd
         SOVERSION ${AMD_VERSION_MAJOR}
         PUBLIC_HEADER "Include/amd.h"
-        WINDOWS_EXPORT_ALL_SYMBOLS ON )
+        WINDOWS_EXPORT_ALL_SYMBOLS ON
+        LINKER_LANGUAGE C )
 
     if ( ${CMAKE_VERSION} VERSION_GREATER_EQUAL "3.25" )
         set_target_properties ( AMD PROPERTIES EXPORT_NO_SYSTEM ON )
@@ -111,7 +112,8 @@ if ( BUILD_STATIC_LIBS )
         C_STANDARD 11
         C_STANDARD_REQUIRED ON
         OUTPUT_NAME amd
-        PUBLIC_HEADER "Include/amd.h" )
+        PUBLIC_HEADER "Include/amd.h"
+        LINKER_LANGUAGE C )
 
     if ( MSVC OR ("${CMAKE_C_SIMULATE_ID}" STREQUAL "MSVC") )
         set_target_properties ( AMD_static PROPERTIES

--- a/CHOLMOD/CMakeLists.txt
+++ b/CHOLMOD/CMakeLists.txt
@@ -383,7 +383,8 @@ if ( BUILD_SHARED_LIBS )
         OUTPUT_NAME cholmod
         SOVERSION ${CHOLMOD_VERSION_MAJOR}
         PUBLIC_HEADER "Include/cholmod.h"
-        WINDOWS_EXPORT_ALL_SYMBOLS ON )
+        WINDOWS_EXPORT_ALL_SYMBOLS ON
+        LINKER_LANGUAGE C )
 
     if ( CHOLMOD_HAS_CUDA )
         set_target_properties ( CHOLMOD PROPERTIES CUDA_SEPARABLE_COMPILATION ON )
@@ -410,7 +411,8 @@ if ( BUILD_STATIC_LIBS )
         C_STANDARD 11
         C_STANDARD_REQUIRED ON
         OUTPUT_NAME cholmod
-        PUBLIC_HEADER "Include/cholmod.h" )
+        PUBLIC_HEADER "Include/cholmod.h"
+        LINKER_LANGUAGE C )
 
     if ( MSVC OR ("${CMAKE_C_SIMULATE_ID}" STREQUAL "MSVC") )
         set_target_properties ( CHOLMOD_static PROPERTIES

